scripts/run_all.py

```diff
@@ -4,86 +4,239 @@
+    def __init__(self, close: bool = False, web: bool = False):
+        self.port = get_free_port()
+        self.should_close = close
+        self.web = web
+        self.web_viewer_port = get_free_port()
+        self.ws_server_port = get_free_port()
+        self.process = None
+
+    def close(self) -> None:
+        if self.process is not None:
+            self.process.kill()
+        self.process = None
+
+    def start(self) -> "Viewer":
+        args = ["--port", str(self.port)]
+        if self.web:
+            args += [
+                "--web-viewer",
+                "--web-viewer-port",
+                str(self.web_viewer_port),
+                "--ws-server-port",
+                str(self.ws_server_port),
+            ]
+
+        self.process = subprocess.Popen(
+            ["cargo", "run", "-p", "rerun", "--all-features", "--"] + args,
+            stdout=subprocess.PIPE,
+        )
+        time.sleep(1)  # give it a moment to start
+        return self
+
+    def __enter__(self) -> "Viewer":
+        self.start()
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc: Optional[BaseException],
+        traceback: Optional[TracebackType],
+    ) -> None:
+        if self.process is not None and self.should_close:
+            self.process.kill()
+
+
+def run_sdk_build() -> None:
+    returncode = subprocess.Popen(
+        ["maturin", "develop", "--manifest-path", "rerun_py/Cargo.toml", '--extras="tests"'],
+    ).wait()
+    assert returncode == 0, f"process exited with error code {returncode}"
 
 
-def run_build() -> None:
-    process = subprocess.Popen(
-        ["maturin", "develop", "--manifest-path", "rerun_py/Cargo.toml", '--extras="tests"'],
-    )
-    returncode = process.wait()
+def run_viewer_build() -> None:
+    returncode = subprocess.Popen(["cargo", "build", "-p", "rerun", "--all-features"]).wait()
     assert returncode == 0, f"process exited with error code {returncode}"
 
 
+def run_web(examples: List[str], separate: bool) -> None:
+    if not separate:
+        with Viewer(close=True, web=True) as viewer:
+            for path in examples:
+                run_py_example(path, viewer_port=viewer.port)
+        return
+
+    cleanup: List[Tuple[Any, Any]] = []
+    # start all examples in parallel
+    for path in examples:
+        # each example gets its own viewer
+        viewer = Viewer(web=True).start()
+        example = run_py_example(path, viewer_port=viewer.port, wait=False)
+        cleanup.append((viewer, example))
+
+    # wait for all processes to finish, and close the viewers
+    for pair in cleanup:
+        viewer, example = pair
+        example.wait()
+        viewer.close()
+
+
+def run_save(examples: List[str]) -> None:
+    with Viewer(close=True) as viewer:  # ephemeral viewer that exists only while saving
+        for example in examples:
+            run_py_example(example, viewer_port=viewer.port, save="out.rrd")
```

Comment:
I don't think `--save` should start a viewer at all. It should just run each example with `--save out.rrd`

Currently running `./scripts/run_all.py --fast --save` produces some weird situation where examples are actually loaded in the viewer (and _also_ saved? I'm unsure)

Reply:
I agree, this came out of a misunderstanding of how the rerun viewer works. I thought it always opened the UI. Not sure how I came to that conclusion, though!

Changed it to run the examples without a viewer.